### PR TITLE
Add ability to set sublayer between vendor and application

### PR DESCRIPTION
**Pull request by devanych**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #95

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -216,6 +218,70 @@ $config = new Config(
 $web = $config->get('web');
 ```
 
+### `over-vendor-layer`
+
+The `over-vendor-layer` option adds a sublayer to the vendor, which allocates packages that will override
+the vendor's default configurations. This sublayer is located between the vendor and application layers.
+
+This can be useful if you need to redefine default configurations even before the application layer. To do this,
+you need to have to create your own package and store configurations in it that override the default ones:
```
Comment (samdark) [inline README.md:227]:
```suggestion
you need to create your own package with configurations meant to override the default ones:
```

File: README.md [hunk 2]
```diff
@@ -216,6 +218,70 @@ $config = new Config(
...
+    "config-plugin-options": {
+        "over-vendor-layer": [
+            "vendor-name/package-1",
+            "vendor-name/package-2"
+        ]
+    }
+}
+```
+
+If there are a lot of packages, so as not to prescribe all the packages, you can use a wildcard pattern:
```
Comment (samdark) [inline README.md:268]:
```suggestion
You can use wildcard pattern if there are too many packages:
```

File: README.md [hunk 3]
```diff
@@ -216,6 +218,70 @@ $config = new Config(
...
+    "config-plugin-options": {
+        "over-vendor-layer": [
+            "vendor-1/*",
+            "vendor-2/config-*"
+        ]
+    }
+}
+```
+
+For more information about the substitution template, see the [yiisoft/strings](https://github.com/yiisoft/strings).
```
Comment (samdark) [inline README.md:281]:
```suggestion
For more information about the wildcard syntax, see the [yiisoft/strings](https://github.com/yiisoft/strings).
```

File: README.md [hunk 4]
```diff
@@ -216,6 +218,70 @@ $config = new Config(
...
+
+> Please note that in this sublayer keys with the same names are not allowed in the same way as in the main layers.
```
Comment (samdark) [inline README.md:283]:
```suggestion
> Please note that in this sublayer keys with the same names are not allowed similar to other layers.
```

File: README.md [hunk 5]
```diff
@@ -216,6 +218,70 @@ $config = new Config(
 $web = $config->get('web');
 ```
 
+### `over-vendor-layer`
```
Comment (samdark) [inline README.md:221]: I think explicit name explains it better:

```suggestion
### `vendor-override-layer`
```

File: src/Composer/ProcessHelper.php [hunk 6]
```diff
@@ -45,16 +52,53 @@ public function __construct(Composer $composer)
...
+     * @psalm-return array<string, CompletePackage>
+     */
+    public function getPackages(): array
+    {
+        return $this->packages;
     }
 
     /**
-     * Builds and returns packages.
+     * Returns vendor packages without packages from the sublayer over the vendor.
```
Comment (samdark) [inline src/Composer/ProcessHelper.php:69]:
```suggestion
     * Returns vendor packages without packages from the vendor override sublayer.
```

File: src/Composer/ProcessHelper.php [hunk 7]
```diff
@@ -45,16 +52,53 @@ public function __construct(Composer $composer)
...
+     * @psalm-return array<string, CompletePackage>
      */
-    public function buildPackages(): array
+    public function getVendorPackages(): array
     {
-        return (new PackagesListBuilder($this->composer))->build();
+        $vendorPackages = [];
+
+        foreach ($this->packages as $name => $package) {
+            if (!$this->isOverVendorPackage($name)) {
```
Comment (samdark) [inline src/Composer/ProcessHelper.php:78]:
```suggestion
            if (!$this->isVendorOverridePackage($name)) {
```

File: src/Options.php [hunk 8]
```diff
@@ -19,8 +19,10 @@ final class Options
     public const DEFAULT_VENDOR_DIRECTORY = 'vendor';
     public const DEFAULT_ENVIRONMENT = '/';
     public const ROOT_PACKAGE_NAME = '/';
+    public const OVER_VENDOR_PACKAGE_NAME = '//';
```
Comment (samdark) [inline src/Options.php:22]:
```suggestion
    public const VENDOR_OVERRIDE_PACKAGE_NAME = '//';
```

**Review by samdark — COMMENTED** [review]
Looks good conceptually. Need approval from @viktorprogger that his case is solved this way.

**Comment by devanych** [comment]
> Looks good conceptually. Need approval from @viktorprogger that his case is solved this way.

I didn't under the case of @viktorprogger.
I did it under the case @thenotsoft. :blush:

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**